From 97997752d8b5d34bb4821ea9de4dd0d1952bec44 Mon Sep 17 00:00:00 2001
From: micbis <michele.bisogno.ct@renesas.com>
Date: Mon, 10 May 2021 10:30:15 +0200
Subject: [PATCH 2/3] target/arm_adi_v5: Fixes for Renesas RZ/G2L

This fix to mem_ap functions is required to support RZ/G2L. The
problem is related to the back to back CSW write and TAR, that
causes a WAIT response from DAP. Folling commands, even if
repeated later in the correct order cause problems to the DAP
itself. Solved by ensuring that as soon as a WAIT is received
the commands are repeated and others are not sent.

Change-Id: Ia0dcbdd8b975083e2462976ccdca26875a1add31
Signed-off-by: micbis <michele.bisogno.ct@renesas.com>
---
 src/target/arm_adi_v5.c | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/src/target/arm_adi_v5.c b/src/target/arm_adi_v5.c
index 8f5ad59c3..f2bcf828e 100644
--- a/src/target/arm_adi_v5.c
+++ b/src/target/arm_adi_v5.c
@@ -379,6 +379,13 @@ static int mem_ap_write(struct adiv5_ap *ap, const uint8_t *buffer, uint32_t siz
 		if (retval != ERROR_OK)
 			break;
 
+		/* Workaround to prevent DAP from locking due to CSW/TAR back to back
+		 * writes that cause a WAIT response from the DAP in some devices
+		 * (e.g. Renesas RZ/G2L) */
+		retval = dap_run(dap);
+		if (retval != ERROR_OK)
+			return retval;
+
 		retval = mem_ap_setup_tar(ap, address ^ addr_xor);
 		if (retval != ERROR_OK)
 			return retval;
@@ -512,6 +519,13 @@ static int mem_ap_read(struct adiv5_ap *ap, uint8_t *buffer, uint32_t size, uint
 		if (retval != ERROR_OK)
 			break;
 
+		/* Workaround to prevent DAP from locking due to CSW/TAR back to back
+		 * writes that cause a WAIT response from the DAP in some devices
+		 * (e.g. Renesas RZ/G2L) */
+		retval = dap_run(dap);
+		if (retval != ERROR_OK)
+			return retval;
+
 		retval = mem_ap_setup_tar(ap, address);
 		if (retval != ERROR_OK)
 			break;
@@ -767,6 +781,13 @@ int mem_ap_init(struct adiv5_ap *ap)
 	if (retval != ERROR_OK)
 		return retval;
 
+	/* Workaround to prevent DAP from locking due to CSW/TAR back to back
+	 * writes that cause a WAIT response from the DAP in some devices
+	 * (e.g. Renesas RZ/G2L) */
+	retval = dap_run(dap);
+	if (retval != ERROR_OK)
+		return retval;
+
 	retval = dap_queue_ap_read(ap, MEM_AP_REG_CSW, &csw);
 	if (retval != ERROR_OK)
 		return retval;
-- 
2.25.1

